include/aie/detail/emulation.hpp

```diff
@@ -11,16 +11,214 @@
+  int size_x;
+  int size_y;
+  std::atomic<void*> services = nullptr;
+  fiber_pool pool{1, fiber_pool::sched::round_robin, false};
+
+  tile_info& get_tile_info(position pos) {
+    assert(pos.x < size_x && pos.y < size_y);
+    return tiles[pos.x * size_y + pos.y];
+  }
+
+  device_impl() = default;
+
+  device_impl(int x, int y) {
+    size_x = x;
+    size_y = y;
+    tiles.resize(x * y);
+  }
+
+  void add_storage(position pos, void* storage) { get_tile_info(pos).mem = storage; }
+
+  void* get_mem(position pos) { return get_tile_info(pos).mem; }
+
+  template <typename ServiceTy> void wait_all(ServiceTy&&);
+
+  ~device_impl() {
+  }
+};
+
+struct host_tile_impl {
+  position pos;
+  device_impl* dev;
+  fiber_pool::future<void> future_work;
+
+  void init(device_impl& d, position p) {
+    pos = p;
+    dev = &d;
+    d.get_tile_info(p).tile = this;
+  }
+
+  host_lock_impl lock(dir d, int i) {
+    return dev->get_tile_info(pos.on(d)).locks.lock(i);
+  }
+
+  void* get_mem_addr(dir d) { return dev->get_mem(pos.on(d)); }
+
+  template <typename LambdaTy, typename DeviceTileTy>
+  void execute(LambdaTy l, DeviceTileTy& dt) {
+    /// This is the very simple and easy way to do it. but it only works for 1
+    /// by 1 tiles. because it is fully sychronous so no tile can concurently
+    /// execute.
+    future_work = dev->pool.submit([=] mutable { l(dt); });
+  }
+
+  template <typename AccTy> void register_accessor(const AccTy&) {
+    /// This is only needed if we start supporting async execution like
+    /// SYCL.
+  }
+
+  void notify_has_accessed_mem(void* mem, std::size_t size) {
+    /// This is only needed on hardware.
+  }
+
+  host_lock_impl lock(int i) { return lock(aie::dir::self, i); }
+};
+
+template <typename ServiceTy> void device_impl::wait_all(ServiceTy&& s) {
+  auto serv = std::move(s);
+  services.store(&serv);
+
+  for (auto& tile : tiles)
+    tile.tile->future_work.wait();
 }
 
+struct device_tile_impl {
+  position pos;
+  device_impl* dev;
+
+  void init(device_impl& d, position p) {
+    pos = p;
+    dev = &d;
+  }
+
+  device_lock_impl lock(position pos, dir d, int i) {
+    return dev->get_tile_info(pos.on(d)).locks.lock(i);
+  }
+
+  template <dir d> void* get_mem_addr() { return dev->get_mem(pos.on(d)); }
+
+  void stream_write16(const char* ptr, int stream_dix) { TRISYCL_FALLBACK; }
+
+  void stream_read16(char* ptr, int stream_dix) { TRISYCL_FALLBACK; }
+
+  void cascade_write48(const char* ptr) { TRISYCL_FALLBACK; }
+
+  void cascade_read48(const char* ptr) { TRISYCL_FALLBACK; }
+
+  int x_coord() { return pos.x; }
+
+  int y_coord() { return pos.y; }
+
+  template <typename T, typename ServiceTy, typename ServiceStorageTy>
+  typename service_info<T>::ret_t
+  perform_service(typename service_info<T>::data_t data, bool) {
+    ServiceStorageTy* serv = nullptr;
+
+    /// wait for the service to become available
+    while (!(serv = (ServiceStorageTy*)dev->services.load()))
```

Comment:
This should use a synchronization primitive instead of an active polling.
Just a `boost::fiber::condition_variable` could work.